Sources/AsyncHTTPClient/AsyncAwait/HTTPClientRequest.swift

```diff
@@ -31,4 +32,96 @@ struct HTTPClientRequest {
     }
 }
 
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+extension HTTPClientRequest {
+    struct Body {
+        internal enum Mode {
+            case asyncSequence(length: Int?, (ByteBufferAllocator) async throws -> ByteBuffer?)
+            case sequence(length: Int?, (ByteBufferAllocator) -> ByteBuffer)
+            case byteBuffer(ByteBuffer)
+        }
+
+        var mode: Mode
+
+        private init(_ mode: Mode) {
+            self.mode = mode
+        }
+    }
+}
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+extension HTTPClientRequest.Body {
+    static func byteBuffer(_ byteBuffer: ByteBuffer) -> Self {
+        self.init(.byteBuffer(byteBuffer))
+    }
+
+    static func bytes<Bytes>(
+        length: Int? = nil,
+        _ bytes: Bytes
+    ) -> Self where Bytes: Sequence, Bytes.Element == UInt8 {
+        self.init(.sequence(length: length) { allocator in
+            if let buffer = bytes.withContiguousStorageIfAvailable({ allocator.buffer(bytes: $0) }) {
+                // fastpath
+                return buffer
+            }
+            // potentially really slow path
+            return allocator.buffer(bytes: bytes)
+        })
+    }
+
+    static func bytes<Bytes>(
+        _ bytes: Bytes
+    ) -> Self where Bytes: RandomAccessCollection, Bytes.Element == UInt8 {
+        self.init(.sequence(length: bytes.count) { allocator in
+            if let buffer = bytes.withContiguousStorageIfAvailable({ allocator.buffer(bytes: $0) }) {
+                // fastpath
+                return buffer
+            }
+            // potentially really slow path
+            return allocator.buffer(bytes: bytes)
+        })
+    }
+
+    /// This method should never be used and was always deprecated.
+    /// The whole purpose of this overload is to prevent users from providing a redundant length if `Bytes` conforms to
+    /// `RandomAccessCollection` because it already provide a property `count` to get the length in O(**1**).
+    /// - Note: `length` is ignored in favour of `bytes.count`
+    @available(*, deprecated, message: "no need to manually specify `length` because we automatically use `bytes.count` as the `length`")
+    static func bytes<Bytes>(
+        length: Int,
+        _ collection: Bytes
+    ) -> Self where Bytes: RandomAccessCollection, Bytes.Element == UInt8 {
+        return .bytes(collection)
+    }
```

Comment:
This is an interesting idea... It pushes people into using the right thing. However I wonder if it is a good idea to introduce new API, that we deprecate from the get go. @Lukasa wdyt?

Reply:
I'd be more inclined to consider marking the method `@available(*, unavailable)` and then not giving it an implementation beyond `fatalError`

Reply:
I have tried `@available(*, unavailable)` but then the swift compiler will just choose the overload for Sequence:
https://github.com/swift-server/async-http-client/blob/8e4b2b4267029978821737c42e920dfd6b2fe4d3/Sources/AsyncHTTPClient/AsyncAwait/HTTPClientRequest.swift#L58-L70
Even adding `@_disfavoredOverload` to this overload doesn't help.

I think what we could also do is marking the method as deprecated and just adding a fatalError with a useful message.